nomos-mix/core/src/conn_maintenance.rs

```diff
@@ -0,0 +1,410 @@
+                    } else if meter.is_unhealthy(settings) {
+                        tracing::warn!("Detected an unhealthy peer: {:?}", peer);
+                        unhealthy_peers.insert(peer.clone());
+                    }
+                }
+            });
+        }
+
+        assert!(malicious_peers.is_disjoint(&unhealthy_peers));
+        (malicious_peers, unhealthy_peers)
+    }
+
+    /// Reset monitors for all connected peers to be monitored in the next time window.
+    /// To avoid false detection, we only monitor peers that were already connected
+    /// before the time window started.
+    fn reset_monitors(&mut self) {
+        if let Some(monitors) = self.monitors.as_mut() {
+            *monitors = self
+                .connected_peers
+                .iter()
+                .cloned()
+                .map(|peer| (peer, ConnectionMonitor::new()))
+                .collect();
+        }
+    }
+
+    /// Adjust the number of new connections to establish in order to not exceed `max_peering_degree`.
+    fn adjust_num_to_connect(&self, num_to_connect: usize, num_to_close: usize) -> usize {
+        let max_peering_degree = self.settings.max_peering_degree;
+        let num_peers_after_close = self.connected_peers.len() - num_to_close;
+        if num_peers_after_close + num_to_connect > max_peering_degree {
+            let new_num_to_connect = max_peering_degree - num_peers_after_close;
+            tracing::warn!(
+                "Cannot establish {} new connections due to max_peering_degree:{}. Instead, establishing {} connections",
+                num_to_connect, max_peering_degree, new_num_to_connect
+            );
+            new_num_to_connect
+        } else {
+            num_to_connect
+        }
+    }
+}
+
+/// Meter to count the number of effective and drop messages sent by a peer
+#[derive(Debug)]
+struct ConnectionMonitor {
+    effective_messages: usize,
+    drop_messages: usize,
+}
+
+#[derive(Clone, Copy, Debug, Serialize, Deserialize)]
+pub struct ConnectionMonitorSettings {
+    /// Time interval to measure/evaluate the number of messages sent by each peer.
+    pub time_window: Duration,
+    /// The number of effective (data or cover) messages that a peer is expected to send in a given time window.
+    /// If the measured count is greater than (expected * (1 + tolerance)), the peer is considered malicious.
+    /// If the measured count is less than (expected * (1 - tolerance)), the peer is considered unhealthy.
+    pub expected_effective_messages: f32,
+    pub effective_message_tolerance: f32,
+    /// The number of drop messages that a peer is expected to send in a given time window.
+    /// If the measured count is greater than (expected * (1 + tolerance)), the peer is considered malicious.
+    /// If the measured count is less than (expected * (1 - tolerance)), the peer is considered unhealthy.
+    pub expected_drop_messages: f32,
+    pub drop_message_tolerance: f32,
```

Comment:
Btw, I guess this values are nomalized. It would probably be a good idea to use fixed size precision: https://docs.rs/fixed/latest/fixed/

Reply:
Oh, thanks. I updated code by using fixed::U57F7 because I think it's good enough to have the max 0.99 for the fractional part. Also, I updated the code to handle overflow of U57F7 properly when incrementing counters.
I'm merging this PR now since I think the updated code is safe.